src/structure/pdb.jl

```diff
@@ -30,26 +49,315 @@ end
+    end 
+    # Check if the PDB file is marked as obsolete
+    if obsolete
+        # Set the download path to obsolete directory inside the "pdb_dir"
+        pdb_dir = joinpath(pdb_dir,"obsolete")
+    end
+    # Check and create directory if it does not exists in filesystem
+    if !isdir(pdb_dir) 
+        info("Creating directory : $pdb_dir")
+        mkpath(pdb_dir)
     end
-    if ba_number == 0
-        download("http://www.rcsb.org/pdb/files/$pdbid.pdb", out_filepath)
+    archivefilepath = tempname()
+    # The PDB file name will end with the ba_number when downloading biological assembly
+    if ba_number==0
+        pdbpath = joinpath(pdb_dir,"$pdbid"*pdbextension[file_format])
     else
-        # Will download error page if ba_number is too high
-        download("http://www.rcsb.org/pdb/files/$pdbid.pdb$ba_number", out_filepath)
+        pdbpath = joinpath(pdb_dir,"$pdbid"*pdbextension[file_format]*"$ba_number")
     end
+    # Download the PDB file only if it does not exist in the "pdb_dir" and when "overwrite" is true
+    if ispath(pdbpath) && !overwrite
+        info("PDB Exists : $pdbid")
+    else
+        info("Downloading PDB : $pdbid")
+        if ba_number == 0            
+            if file_format == PDB || file_format == PDBXML || file_format == mmCIF
+                download("http://files.rcsb.org/download/$pdbid"*pdbextension[file_format]*".gz", archivefilepath)
+            elseif file_format == MMTF
+                # MMTF is downloaded in uncompressed form, thus directly stored in pdbpath
+                download("http://mmtf.rcsb.org/v1.0/full/$pdbid", pdbpath)
+            else
+                warn("Invalid PDB file format!")
+            end
+        else            
+            if file_format == PDB
+                download("http://files.rcsb.org/download/$pdbid"*pdbextension[file_format]*"$ba_number.gz",archivefilepath)
+            elseif file_format == mmCIF
+                download("http://files.rcsb.org/download/$pdbid-assembly$ba_number"*pdbextension[file_format]*".gz", archivefilepath)
+            else
+                warn("Biological Assembly is available only in PDB and mmCIF formats!")
+            end
+        end
+        # Verify if the compressed PDB file is downloaded properly and extract it. For MMTF no extraction is needed
+        if ispath(archivefilepath) && filesize(archivefilepath) > 0 && file_format != MMTF           
+            input = open(archivefilepath) |> ZlibInflateInputStream
+            open(pdbpath,"w") do output
+                for line in eachline(input)
+                    println(output, line)
+                end
+            end
+            close(input)
+        end
+        # Verify if the PDB file is downloaded without any error
+        if !ispath(pdbpath) || filesize(pdbpath)==0
+            throw(ErrorException("Error downloading PDB : $pdbid"))
+        end
+    end
+    rm(archivefilepath)
```

Comment:
Interesting, it doesn't do that on my machine. I wonder if that is system specific. For example if I attempt to overwrite when overwrite is false I get

```
INFO: PDB Exists : 1AKE
ERROR: unlink: no such file or directory (ENOENT)
 in unlink(::String) at /Applications/Julia-0.5.app/Contents/Resources/julia/lib/julia/sys.dylib:?
 in #rm#7(::Bool, ::Bool, ::Function, ::String) at /Applications/Julia-0.5.app/Contents/Resources/julia/lib/julia/sys.dylib:?
 in #downloadpdb#1(::String, ::Type{T}, ::Bool, ::Bool, ::Int64, ::Function, ::String) at ./REPL[1]:61
 in (::#kw##downloadpdb)(::Array{Any,1}, ::#downloadpdb, ::String) at ./<missing>:0
```